dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/h12/grpc/GrpcCompositeCodec.java

```diff
@@ -16,123 +16,103 @@
-            return protobufHttpMessageCodec.decode(inputStream, targetType, charset);
+        try {
+            byte[] data = StreamUtils.readBytes(inputStream);
+            return packableMethod.parseRequest(data);
+        } catch (Exception e) {
+            throw new DecodeException(e);
         }
-        return wrapperHttpMessageCodec.decode(inputStream, targetType, charset);
     }
 
     @Override
     public Object[] decode(InputStream inputStream, Class<?>[] targetTypes, Charset charset) throws DecodeException {
-        if (targetTypes.length > 1) {
-            return wrapperHttpMessageCodec.decode(inputStream, targetTypes, charset);
-        }
-        return HttpMessageCodec.super.decode(inputStream, targetTypes, charset);
-    }
-
-    private static void writeLength(OutputStream outputStream, int length) {
-        try {
-            outputStream.write(((length >> 24) & 0xFF));
-            outputStream.write(((length >> 16) & 0xFF));
-            outputStream.write(((length >> 8) & 0xFF));
-            outputStream.write((length & 0xFF));
-        } catch (IOException e) {
-            throw new EncodeException(e);
+        Object message = decode(inputStream, ArrayUtils.isEmpty(targetTypes) ? null : targetTypes[0], charset);
+        if (message instanceof Object[]) {
+            return (Object[]) message;
         }
+        return new Object[] {message};
     }
 
     @Override
     public MediaType mediaType() {
         return MediaType.APPLICATION_GRPC;
     }
 
-    private static boolean isProtobuf(Object data) {
-        if (data == null) {
-            return false;
-        }
-        return isProtoClass(data.getClass());
-    }
-
-    private static boolean isProtoClass(Class<?> clazz) {
-        while (clazz != Object.class && clazz != null) {
-            Class<?>[] interfaces = clazz.getInterfaces();
-            if (interfaces.length > 0) {
-                for (Class<?> clazzInterface : interfaces) {
-                    if (PROTOBUF_MESSAGE_CLASS_NAME.equalsIgnoreCase(clazzInterface.getName())) {
-                        return true;
-                    }
-                }
-            }
-            clazz = clazz.getSuperclass();
-        }
-        return false;
-    }
-
-    /**
-     * lazy init protobuf class
-     */
-    private static class ProtobufWriter {
-
-        private static void write(HttpMessageCodec codec, OutputStream outputStream, Object data) {
-            int serializedSize = ((Message) data).getSerializedSize();
-            // write length
-            writeLength(outputStream, serializedSize);
-            codec.encode(outputStream, data);
-        }
+    private void writeLength(OutputStream outputStream, int length) throws IOException {
+        outputStream.write(((length >> 24) & 0xFF));
+        outputStream.write(((length >> 16) & 0xFF));
+        outputStream.write(((length >> 8) & 0xFF));
+        outputStream.write((length & 0xFF));
```

Comment:
**Performance Issue**: Refactoring the codec to use a more streamlined approach for handling encoding and decoding, focusing on performance and maintainability. <br> **Fix**: Refactor the codec implementation to focus on performance by using a more efficient method for encoding and decoding, ensuring thread safety and reducing complexity. <br> **Code Suggestion**: 
 ```
 Refactor the codec implementation in GrpcCompositeCodec.java to improve performance and maintainability. Focus on streamlining the approach for handling encoding and decoding. Ensure the new implementation is thread-safe and reduces complexity.
 ```